### PR TITLE
Fix: add collect methods for Difference and Intersection

### PR DESCRIPTION
**Pull request by saschahofmann**
Adds the collect functions for Difference and Intersection

ref #2845

### CODE REVIEW

**Comment by saschahofmann** [comment]
@datapythonista I think the failures are not related to this PR.

Ready to merge IMO

**Review by jreback — CHANGES_REQUESTED** [review]
what tests fail before this change and now work?

**Comment by saschahofmann** [comment]
This fixes untested behaviour https://github.com/ibis-project/ibis/issues/2845#issuecomment-876627886

TLDR:
When you try to "select"=project from a difference or intersection it returns an error because the methods added here aren't defined.

**Comment by jreback** [comment]
ok can u add a test to that effect

**Comment by saschahofmann** [comment]
Sure any suggestions where I should put that?

**Comment by jreback** [comment]
see how collect_Union is tested and mimic that, we test pretty generically for sql backends

**Comment by saschahofmann** [comment]
Ok I added a test to _test_compiler.py_. I think that should do it?

File: ibis/tests/sql/test_compiler.py [hunk 1]
```diff
@@ -2298,6 +2298,41 @@ def test_table_difference(self):
...
+        expected = """SELECT `key`
+FROM (
+  SELECT `string_col` AS `key`, CAST(`float_col` AS double) AS `value`
+  FROM functional_alltypes
+  WHERE `int_col` > 0
+  INTERSECT
+  SELECT `string_col` AS `key`, `double_col` AS `value`
+  FROM functional_alltypes
+  WHERE `int_col` <= 0
+) t0"""
```
Comment (datapythonista) [inline ibis/tests/sql/test_compiler.py:2318]: What do you think about creating a constant with this, with the `INTERSECT` / `EXCEPT` as variable, and in the tests just use `QUERY.format('INTERSECT')`? Besides avoiding the duplication, I think it should make tests more readable.

Comment (saschahofmann) [reply ibis/tests/sql/test_compiler.py]: done. Good idea much nicer now

File: ibis/tests/sql/test_compiler.py [hunk 2]
```diff
@@ -2221,6 +2221,27 @@ def test_endswith(self):
...
+SELECT_QUERY = """SELECT `key`
+FROM (
+  SELECT `string_col` AS `key`, CAST(`float_col` AS double) AS `value`
+  FROM functional_alltypes
+  WHERE `int_col` > 0
+  {}
+  SELECT `string_col` AS `key`, `double_col` AS `value`
+  FROM functional_alltypes
+  WHERE `int_col` <= 0
+) t0"""
```
Comment (datapythonista) [inline ibis/tests/sql/test_compiler.py:2242]: Still a significant amount of duplication, wouldn't make sense to use inside the test something like this or similar instead?
```
SELECT `key` FROM ({QUERY.format('INTERSECT')}) t0
```

Comment (saschahofmann) [reply ibis/tests/sql/test_compiler.py:2242]: It's either that or something like this

```python
QUERY = """\
SELECT `string_col` AS `key`, CAST(`float_col` AS double) AS `value`
FROM functional_alltypes
WHERE `int_col` > 0
{}
SELECT `string_col` AS `key`, `double_col` AS `value`
FROM functional_alltypes
WHERE `int_col` <= 0"""

SELECT_QUERY = f"SELECT `key` FROM ({QUERY}) t0"
```

I would just need to some add new the two spaces in front of each new line `QUERY`. Maybe your suggestions is nicer?

Comment (datapythonista) [reply ibis/tests/sql/test_compiler.py:2242]: Ah, true. i think it's still worth, you can use `textwrap.indent`. But up to you, no big deal.

Comment (saschahofmann) [reply ibis/tests/sql/test_compiler.py]: Haha should have known there is a package for that. Done!

File: docs/source/release/index.rst [hunk 3]
```diff
@@ -11,7 +11,8 @@ Release Notes
 
    These release notes are for versions of ibis **1.0 and later**. Release
    notes for pre-1.0 versions of ibis can be found at :doc:`release-pre-1.0`
-
+   
+* :bug:`2845` Fix projection on differences and intersections
```
Comment (jreback) [inline docs/source/release/index.rst:15]: can you say for the SQL backends.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, small comment

File: docs/source/release/index.rst (hunk 3 above)
Comment (saschahofmann) [reply docs/source/release/index.rst]: sure thing done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @saschahofmann

**Pull request merged**